js/src/sheet.js

```diff
@@ -51,93 +49,99 @@ var SheetModel = widgets.DOMWidgetModel.extend({
     },
-    on_change_cells: function() {
+    on_change_cells: function () {
         console.log('change cells');
         this._updating_grid = true;
         try {
             var previous_cells = this.previous('cells');
             var cells = this.get('cells');
-            for(var i = 0; i < cells.length; i++) {
+            for (var i = 0; i < cells.length; i++) {
                 var cell = cells[i];
-                if(!contains(previous_cells, cell)) {
+                if (!contains(previous_cells, cell)) {
                     console.log('adding cell', cell);
                     this.cell_bind(cell);
                 }
             }
-            this.cells_to_grid()
-            //this.save_changes();
+            this.cells_to_grid();
+            // this.save_changes();
         } finally {
             this._updating_grid = false;
         }
-        this.grid_to_cell()
+        this.grid_to_cell();
     },
-    cell_bind: function(cell) {
-        //this.cell_to_grid(cell, false);
-        cell.on('change:value change:style change:type change:renderer change:read_only change:choice change:format', function() {
-            //this.cell_to_grid(cell, true);
-            this.cells_to_grid()
+    cell_bind: function (cell) {
+        // this.cell_to_grid(cell, false);
+        cell.on('change:value change:style change:type change:renderer change:read_only change:choice change:format', function () {
+            // this.cell_to_grid(cell, true);
+            this.cells_to_grid();
         }, this);
     },
-    cells_to_grid: function() {
+    cells_to_grid: function () {
         var data = cloneDeep(this.get('data'));
         each(this.get('cells'), (cell) => {
-            this._cell_data_to_grid(cell, data)
-        })
+            this._cell_data_to_grid(cell, data);
+        });
         this.set('data', data);
-        this.save_changes()
+        this.save_changes();
     },
-    cell_to_grid: function(cell, save) {
+    cell_to_grid: function (cell, save) {
         console.log('cell to grid', cell);
         var data = cloneDeep(this.get('data'));
-        this._cell_data_to_grid(cell, data)
+        this._cell_data_to_grid(cell, data);
         this.set('data', data);
-        if(save) {
+        if (save) {
             this.save_changes();
         }
     },
-    _cell_data_to_grid: function(cell, data) {
-        var value = cell.get('value');
-        if(!value)
-            return
-        for(var i = cell.get('row_start'); i <= cell.get('row_end'); i++) {
-            for(var j = cell.get('column_start'); j <= cell.get('column_end'); j++) {
+    _cell_data_to_grid: function (cell, data) {
+        var val = cell.get('value');
+        if (!val) {
+            return;
+        }
+        for (var i = cell.get('row_start'); i <= cell.get('row_end'); i++) {
+            for (var j = cell.get('column_start'); j <= cell.get('column_end'); j++) {
                 var value = cell.get('value');
```

Comment:
Not sure why, but tests were not passing when removing this line, but `value` is already defined a few lines above... Is `value` supposed to have changed since the last call of `cell.get('value')`?

Reply:
I need to check that locally, looking at the code I cannot figure it out, makes me feel good about writing the tests though :)

Reply:
I can remove like 117 without any issue, can you double check?

Reply:
Sure

Reply:
This command:
```
cd js && npm install && ./node_modules/.bin/karma start --single-run
```
works fine when line 117 is here. But when I remove it I have those errors:
```
HeadlessChrome 73.0.3679 (Linux 0.0.0) sheet range should be reflected in all cells FAILED
	TypeError: Cannot read property '1' of undefined
	    at child._cell_data_to_grid (/tmp/karma-typescript-bundle-24904Dgtbs36tbibs.js:177134:38)
	    at /tmp/karma-typescript-bundle-24904Dgtbs36tbibs.js:177093:20
	    at arrayEach (/tmp/karma-typescript-bundle-24904Dgtbs36tbibs.js:28655:11)
	    at forEach (/tmp/karma-typescript-bundle-24904Dgtbs36tbibs.js:37483:14)
	    at child.cells_to_grid (/tmp/karma-typescript-bundle-24904Dgtbs36tbibs.js:177092:26)
	    at child.on_change_cells (/tmp/karma-typescript-bundle-24904Dgtbs36tbibs.js:177074:18)
	    at triggerEvents (/tmp/karma-typescript-bundle-24904Dgtbs36tbibs.js:25219:57)
	    at triggerApi (/tmp/karma-typescript-bundle-24904Dgtbs36tbibs.js:25204:19)
	    at eventsApi (/tmp/karma-typescript-bundle-24904Dgtbs36tbibs.js:25003:16)
	    at child.Events.trigger (/tmp/karma-typescript-bundle-24904Dgtbs36tbibs.js:25194:5)
LOG: 'change cells'
LOG: 'adding cell', child{cid: 'c30', attributes: Object{squeeze_row: false, squeeze_column: false, row_start: 0, row_end: 1, column_start: 1, column_end: 2, transpose: true, value: [..., ...], _model_module: 'ipysheet', _model_name: 'CellRangeModel', _model_module_version: '~0.2.0', _view_module: '@jupyter-widgets/base', _view_name: null, _view_module_version: '1.1.0', _view_count: null, type: null, name: null, style: Object{}, renderer: null, read_only: false, choice: null, format: '0.[000]'}, _changing: false, _previousAttributes: Object{}, changed: Object{}, id: undefined, _pending: false, widget_manager: DummyManager{comm_target_name: 'jupyter.widget', _models: Object{u-u-i-d: ..., u-u-i-d-cell: ...}, el: <div></div>, library: Object{ipysheet: ...}}, model_id: 'u-u-i-d-cell', views: Object{}, state_change: Promise{}, _closed: false, _state_lock: null, _msg_buffer: null, _msg_buffer_callbacks: null, _pending_msgs: 1, _buffered_state_diff: Object{}, comm: MockComm{_on_msg: function () { ... }, _on_open: null, _on_close: function () { ... }, comm_id: 'mock-comm-id-18'}, _comm_live: true, name: 'CellRangeModel', module: 'ipysheet', _events: Object{comm:close: [...]}}
HeadlessChrome 73.0.3679 (Linux 0.0.0) sheet range tranposed should be reflected in all cells FAILED
	TypeError: Cannot read property '0' of undefined
	    at child._cell_data_to_grid (/tmp/karma-typescript-bundle-24904Dgtbs36tbibs.js:177127:38)
	    at /tmp/karma-typescript-bundle-24904Dgtbs36tbibs.js:177093:20
	    at arrayEach (/tmp/karma-typescript-bundle-24904Dgtbs36tbibs.js:28655:11)
	    at forEach (/tmp/karma-typescript-bundle-24904Dgtbs36tbibs.js:37483:14)
	    at child.cells_to_grid (/tmp/karma-typescript-bundle-24904Dgtbs36tbibs.js:177092:26)
	    at child.on_change_cells (/tmp/karma-typescript-bundle-24904Dgtbs36tbibs.js:177074:18)
	    at triggerEvents (/tmp/karma-typescript-bundle-24904Dgtbs36tbibs.js:25219:57)
	    at triggerApi (/tmp/karma-typescript-bundle-24904Dgtbs36tbibs.js:25204:19)
	    at eventsApi (/tmp/karma-typescript-bundle-24904Dgtbs36tbibs.js:25003:16)
	    at child.Events.trigger (/tmp/karma-typescript-bundle-24904Dgtbs36tbibs.js:25194:5)
LOG: 'change cells'
LOG: 'adding cell', child{cid: 'c32', attributes: Object{squeeze_row: false, squeeze_column: false, row_start: 0, row_end: 1, column_start: 1, column_end: 2, transpose: true, value: [..., ...], style: Object{color: ..., backgrouncColor: ...}, _model_module: 'ipysheet', _model_name: 'CellRangeModel', _model_module_version: '~0.2.0', _view_module: '@jupyter-widgets/base', _view_name: null, _view_module_version: '1.1.0', _view_count: null, type: null, name: null, renderer: null, read_only: false, choice: null, format: '0.[000]'}, _changing: false, _previousAttributes: Object{}, changed: Object{}, id: undefined, _pending: false, widget_manager: DummyManager{comm_target_name: 'jupyter.widget', _models: Object{u-u-i-d: ..., u-u-i-d-cell: ...}, el: <div></div>, library: Object{ipysheet: ...}}, model_id: 'u-u-i-d-cell', views: Object{}, state_change: Promise{}, _closed: false, _state_lock: null, _msg_buffer: null, _msg_buffer_callbacks: null, _pending_msgs: 1, _buffered_state_diff: Object{}, comm: MockComm{_on_msg: function () { ... }, _on_open: null, _on_close: function () { ... }, comm_id: 'mock-comm-id-20'}, _comm_live: true, name: 'CellRangeModel', module: 'ipysheet', _events: Object{comm:close: [...]}}
HeadlessChrome 73.0.3679 (Linux 0.0.0) sheet should use the last style but not overwrite FAILED
	TypeError: Cannot read property '0' of undefined
	    at child._cell_data_to_grid (/tmp/karma-typescript-bundle-24904Dgtbs36tbibs.js:177127:38)
	    at /tmp/karma-typescript-bundle-24904Dgtbs36tbibs.js:177093:20
	    at arrayEach (/tmp/karma-typescript-bundle-24904Dgtbs36tbibs.js:28655:11)
	    at forEach (/tmp/karma-typescript-bundle-24904Dgtbs36tbibs.js:37483:14)
	    at child.cells_to_grid (/tmp/karma-typescript-bundle-24904Dgtbs36tbibs.js:177092:26)
	    at child.on_change_cells (/tmp/karma-typescript-bundle-24904Dgtbs36tbibs.js:177074:18)
	    at triggerEvents (/tmp/karma-typescript-bundle-24904Dgtbs36tbibs.js:25219:57)
	    at triggerApi (/tmp/karma-typescript-bundle-24904Dgtbs36tbibs.js:25204:19)
	    at eventsApi (/tmp/karma-typescript-bundle-24904Dgtbs36tbibs.js:25003:16)
	    at child.Events.trigger (/tmp/karma-typescript-bundle-24904Dgtbs36tbibs.js:25194:5)
HeadlessChrome 73.0.3679 (Linux 0.0.0): Executed 15 of 16 (3 FAILED) (skipped 1) (0.349 secs / 0.247 secs)
```

Reply:
Again, a karma issue, it doesn't recompile on changes, so I had to restart. But if `cell.get('value');` is a list or 'matrix' it will be reassinged the value of value[4][6] for instance, so the next iteration it needs to reassign it to the original value again. Maybe not the most readable code, feel free to change it into something more readable.